### PR TITLE
Shrink `ast::Expr` harder

### PR DESCRIPTION
**Pull request by nnethercote**
r? @ghost

### CODE REVIEW

File: compiler/rustc_ast/src/attr/mod.rs [hunk 1]
```diff
@@ -200,17 +214,6 @@ impl MetaItem {
...
         self.path == name
     }
-
-    /// This is used in case you want the value span instead of the whole attribute. Example:
-    ///
-    /// ```text
-    /// #[doc(alias = "foo")]
-    /// ```
-    ///
-    /// In here, it'll return a span for `"foo"`.
```
Comment (fbstj) [inline compiler/rustc_ast/src/attr/mod.rs:210]: this seems to have been dropped

Comment (nnethercote) [reply compiler/rustc_ast/src/attr/mod.rs:210]: It's intentional.

**Comment by nnethercote** [comment]
Lots of nice performance wins here across all the metrics, hooray.

**Comment by nnethercote** [comment]
I fixed the conflicts.

**Comment by nnethercote** [comment]
@petrochenkov: I have rebased again, should be ready for review.

File: compiler/rustc_ast/src/ast.rs [hunk 2]
```diff
@@ -3077,8 +3092,10 @@ mod size_asserts {
     static_assert_size!(AssocItemKind, 32);
     static_assert_size!(Attribute, 32);
     static_assert_size!(Block, 48);
-    static_assert_size!(Expr, 104);
-    static_assert_size!(ExprKind, 72);
+    #[cfg(not(bootstrap))]
```
Comment (petrochenkov) [inline compiler/rustc_ast/src/ast.rs:3095]: Not sure why this `cfg` is needed.
Aren't all these asserts behave identically with regards to bootstrap changes? Why is only `Expr(Kind)` needs to be configured?

**Comment by petrochenkov** [comment]
The `ast::Lit` changes are somewhat controversial, I need to think about the direction here.

r=me on the last 2 commits if you move them to a separate PR.

File: compiler/rustc_ast/src/ast.rs (hunk 2 above)
Comment (nnethercote) [reply compiler/rustc_ast/src/ast.rs:3095]: #94075 recently landed, changing the layout of some types, and shrinking some of them. The `Fn` size assertion in this file is also conditional, as are some in other files, e.g. [hir.rs](https://github.com/rust-lang/rust/blob/6153d3cbe6abc74fb37e4ebe48cc825484fd6bbf/compiler/rustc_hir/src/hir.rs#L3512-L3550).

**Comment by nnethercote** [comment]
> The `ast::Lit` changes are somewhat controversial, I need to think about the direction here.

Any parts in particular?

Without those changes `Expr` only shrinks to 88 bytes. I already tried doing that in #100444 and the perf results were mediocre. I didn't see clear perf wins until shrinking to 72 bytes.

**Comment by nnethercote** [comment]
Rebased.

The first two commits can be viewed as avoiding redundancy.
- Commit 1 is "Don't store the lit kind twice (and in two slightly different ways)".
- Commit 2 is "Don't store the same span twice."

**Comment by petrochenkov** [comment]
@nnethercote 
>Any parts in particular?

Removing `token_lit: token::Lit` instead of `kind: LitKind`.

>Without those changes Expr only shrinks to 88 bytes. I already tried doing that in https://github.com/rust-lang/rust/pull/100444 and the perf results were mediocre. I didn't see clear perf wins until shrinking to 72 bytes.

The suggestion to split the PR into two is mostly to avoid bitrot/rebasing, it will take some time to me to get to the literal issue.

**Comment by nnethercote** [comment]
> Removing `token_lit: token::Lit` instead of `kind: LitKind`.

I first tried removing the `ast::LitKind` first, as suggested by the comment saying "FIXME: Remove this and only create the semantic representation during lowering to HIR." It was painful, because there are various places where you want the semantic representation and you have to reparse the symbol to get it. So it's very useful to have the semantic representation computed once and stored in the AST. And once that is in place, you need an enum for the different kinds. In contrast, removing the token-level kind is easier because it's a simpler enum with very few arguments, and the actual data is in the symbols which don't need an enum.

**Comment by nnethercote** [comment]
See #101885 for removal of `ast::Lit::kind` as a possible alternative.

**Comment by petrochenkov** [comment]
Oh, and moving the span from `ast::Lit` to the outside is also not controversial.

**Comment by petrochenkov** [comment]
Which parts of this PR are unambiguous improvements - https://github.com/rust-lang/rust/pull/101562/commits/9157f34c58d604dcf05a958c384bb77308bac5dc and https://github.com/rust-lang/rust/pull/101562/commits/d38043a493b64c0cb8f40f904d9225ae6aaef0a6?
In that case it makes sense to land them first.

From https://github.com/rust-lang/rust/pull/102000 it looks like the boxing change is a tradeoff, and only makes sense if we shrink literals enough?
In that case it makes sense to land it last, after the question of literals (https://github.com/rust-lang/rust/pull/101562/commits/ceab0b4c1697b93145c08331005453fdc9c7dd01 vs https://github.com/rust-lang/rust/pull/101885) is resolved.

**Comment by nnethercote** [comment]
In terms of performance, none of the changes in isolation are clear improvements. We need all of them (or the `Lit` changes replaced with #101885) to get `Expr` down to 72 bytes to see clear improvements.

**Comment by nnethercote** [comment]
Now that #102944 has been merged, we can try this out again. Now it only requires two commits to shrink `ast::Expr` from 104 bytes to 72 bytes.

@bors try @rust-timer queue

**Comment by nnethercote** [comment]
The cycles, wall-time, max-rss, and cache-misses results here are pretty good.

r? @petrochenkov

**Comment by petrochenkov** [comment]
The error seems unrelated and spurious, but let's check again.
@bors retry

**Comment by matthiaskrgr** [comment]
cc https://github.com/rust-lang/rust/issues/104563 :(

**Comment by nnethercote** [comment]
Instruction count results are mixed. But the real results are for cycles and wall-time, which show lots of improvements.

@rustbot label: +perf-regression-triaged

**Pull request merged**